### PR TITLE
Fix for #108, CommonMark 0.30 spec compliance, part 2

### PR DESCRIPTION
**Pull request by anderskaplan**
A set of small fixes which improve the CommonMark spec compliance. See the commit messages for details.

### CODE REVIEW

**Comment by pbodnar** [comment]
Great, thank you for another contribution. :)

So it looks like, because spec examples 312 and 313 are already fixed in master now, there will be just 1 failing example left after the fixes provided in here, namely:

```
$ py -m test.specification
example:  146
markdown: '~~~ aa ``` ~~~\nfoo\n~~~\n'
html:     '<pre><code class="language-aa">foo\n</code></pre>\n'
output:   '<p><del>~ aa ``` </del>~\nfoo</p>\n<pre><code></code></pre>\n'
```

... the last challenge. ;)

**Comment by anderskaplan** [comment]
Yes, at first I thought it was caused by the strikethrough token, but that can't be. Because the code fences should already be parsed on block level before the strikethrough can kick in on span level.

File: test/test_block_token.py [hunk 1]
```diff
@@ -377,6 +377,22 @@ def test_footnotes_may_not_have_too_much_leading_space(self):
         self.assertIsInstance(token.children[2], block_token.BlockCode)
         self.assertEqual(token.children[2].children[0].content, "[i-am-block-too]: /foo\n")
 
+    def test_footnotes_angle_bracketed_link_with_space(self):
```
Comment (pbodnar) [inline test/test_block_token.py:379]: I can see you've taken this right from the spec, for simplicity. Yet, for better readability within existing tests, I would suggest to move it after `test_parse_with_title()` and to simplify like this:

```python
    def test_parse_with_space_in_every_part(self):
        lines = ['[foo bar]:\n',
                 '<my url>\n',
                 '\'my title\'\n']
        token = block_token.Document(lines)
        self.assertEqual(token.footnotes, {"foo bar": ("my url", "my title")})
```

(no need to import `core_tokens` with this change then)

File: test/test_block_token.py [hunk 2]
```diff
@@ -377,6 +377,22 @@ def test_footnotes_may_not_have_too_much_leading_space(self):
...
+        lines = ['[Foo bar]:\n',
+                 '<my url>\n',
+                 '\'title\'\n',
+                 '\n',
+                 '[Foo bar]\n']
+        token = block_token.Document(lines)
+        self.assertEqual(token.footnotes, {core_tokens.normalize_label("Foo bar"): ("my url", "title")})
+
+    def test_footnotes_title_separated_from_link_destination(self):
```
Comment (pbodnar) [inline test/test_block_token.py:388]: This is an extra test added & taken from Example 201 in spec, right? As in the case above, I would suggest to simplify it, like this:

```python
    def test_no_footnote_when_title_right_after_link_destination(self):
        lines = ['[foo]: <bar>(baz)\n']
        token = block_token.Document(lines)
        self.assertEqual(token.footnotes, {})
```

File: test/test_span_token.py [hunk 3]
```diff
@@ -107,6 +107,13 @@ def test_parse_children(self):
         child = next(iter(token.children))
         self._test_token(child, 'alt', src='src')
 
+    def test_angle_bracketed_inline_link_with_space(self):
```
Comment (pbodnar) [inline test/test_span_token.py:116]: For the sake of simplicity and consistency, I suggest to shorten this to:

```python
    def test_angle_bracketed_inline_link_with_space(self):
        self._test_parse(span_token.Link, '[link](</my uri>)', 'link',
            target='/my uri', title='')
```

File: mistletoe/core_tokens.py [hunk 4]
```diff
@@ -307,6 +307,12 @@ def match_link_label(string, offset, root=None):
 
 
 def is_link_label(text, root):
+    """
+    Normalize and look up `text` among the footnotes.
```
Comment (pbodnar) [inline mistletoe/core_tokens.py:311]: Nice one, there is never enough of documentation.

I would also suggest to rename this old method to `get_link_label` - that would be more appropriate, given what the function actually does (and seemingly did)?

**Review by pbodnar — CHANGES_REQUESTED** [review]
@anderskaplan, good job, I think all the fixes are correct. 👍 I believe there is room for improving the _unit tests_ though (I'm a big fan of having also the test code simple and consistent, to some extent, of course. ;)) - see my suggestions. Thank you.

**Comment by pbodnar** [comment]
> Yes, at first I thought it was caused by the strikethrough token, but that can't be. Because the code fences should already be parsed on block level before the strikethrough can kick in on span level.

Yeah, it looks like mistletoe cannot handle this correctly (stated above the [Example 146](https://spec.commonmark.org/0.30/#example-146))?:

> [Info strings](https://spec.commonmark.org/0.30/#info-string) for tilde code blocks can contain backticks and tildes...

**Comment by anderskaplan** [comment]
hey @pbodnar thanks for the review! I'll try and find some time to update the tests during next week.

File: test/test_span_token.py [hunk 5]
```diff
@@ -157,6 +157,13 @@ def test_no_children(self):
         with self.assertRaises(AttributeError):
             token.children
 
+    def test_html_entities(self):
```
[the remark this answers is not in the document]
Comment (anderskaplan) [reply test/test_span_token.py:160]: ok, agree. Adding a positive test for valid html entities as well.

File: test/test_span_token.py [hunk 6]
```diff
@@ -170,3 +171,18 @@ def test_contains(self):
         self.assertTrue('emphasis' in token)
         self.assertFalse('foo' in token)
 
+
+class TestHTMLSpan(unittest.TestCase):
+    def setUp(self):
+        span_token.add_token(span_token.HTMLSpan)
+        self.addCleanup(span_token.reset_tokens)
+
+    def test_html_span(self):
```
[the remark this answers is not in the document]
Comment (anderskaplan) [reply test/test_span_token.py:187]: ok

File: mistletoe/core_tokens.py [hunk 7]
```diff
@@ -438,10 +438,16 @@ def remove(self, n, left=True):
...
-                    and (self.number + other.number) % 3 == 0)
-
+        if self.type[0] != other.type[0]:
+            return False
+        if self.open and self.close or other.open and other.close:
+            # if either of the delimiters can both open and close emphasis, then additional
+            # restrictions apply: the sum of the lengths of the delimiter runs
+            # containing the opening and closing delimiters must not be a multiple of 3
+            # unless both lengths are multiples of 3.
+            return ((self.number + other.number) % 3 != 0
```
[the remark this answers is not in the document]
Comment (anderskaplan) [reply mistletoe/core_tokens.py:448]: you're right, that's a more precise description than just "modifying the logic".

@pbodnar I can change the commit message, but doing so will be a rebase and therefore mess with the review comments. Do you think it's worth it?

File: test/test_span_token.py [hunk 8]
```diff
@@ -54,6 +54,11 @@ def test_emphasis_for_asterisk_without_punctuation(self):
         self._test_token(next(tokens), 'an example without', children=True)
         self._test_token(next(tokens), 'punctuation', children=False)
 
+    def test_emphasis_when_both_delimiter_run_lengths_are_multiples_of_3(self):
```
[the remark this answers is not in the document]
Comment (anderskaplan) [reply test/test_span_token.py:57]: sure

File: test/test_block_token.py (hunk 1 above)
Comment (anderskaplan) [reply test/test_block_token.py:379]: ok

File: test/test_block_token.py (hunk 2 above)
Comment (anderskaplan) [reply test/test_block_token.py:388]: ok. added a positive test, too.

File: test/test_span_token.py (hunk 3 above)
Comment (anderskaplan) [reply test/test_span_token.py:116]: yes, that's nicer.

File: mistletoe/core_tokens.py (hunk 4 above)
Comment (anderskaplan) [reply mistletoe/core_tokens.py:311]: agree

File: mistletoe/core_tokens.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (anderskaplan) [reply mistletoe/core_tokens.py:448]: well whatever, I'll update the commit message and be done with it :)

**Comment by anderskaplan** [comment]
all done! the last failing test case will be in a PR of its own.

File: mistletoe/core_tokens.py (hunk 4 above)
Comment (pbodnar) [reply mistletoe/core_tokens.py:311]: thanks for the rename :)

File: test/test_span_token.py (hunk 3 above)
Comment (pbodnar) [reply test/test_span_token.py]: thanks

File: test/test_block_token.py (hunk 2 above)
Comment (pbodnar) [reply test/test_block_token.py]: so now it is `test_parse_title_must_be_separated_from_link_destination()` - OK

File: test/test_block_token.py (hunk 1 above)
Comment (pbodnar) [reply test/test_block_token.py]: thanks

File: test/test_span_token.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pbodnar) [reply test/test_span_token.py]: great :)

File: test/test_span_token.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pbodnar) [reply test/test_span_token.py]: thanks

File: mistletoe/core_tokens.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pbodnar) [reply mistletoe/core_tokens.py:448]: looks good, thanks

File: test/test_span_token.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pbodnar) [reply test/test_span_token.py]: done

**Review by pbodnar — COMMENTED** [review]
@anderskaplan, thanks for incorporating all the suggestions, I think it really is all ready for merge now. :)

Yet, would you mind doing a final interactive rebase and squashing?

1) I think it is fine to keep the "fix example X" commits separated. Also b160a88e0e1a6e24ed0c0699373e1abcec874514 could probably stand on its own.
2) I would just remove `#` before the example numbers because of the GitHub automatic issue links and add ` #108` to the end of the summary line. 
3) The commits on fine-tuning unit tests could be either squashed into their corresponding "feature commits" (too much work though?), or squashed into one?

**Comment by anderskaplan** [comment]
Updated the commit messages and squashed the unit test updates according to review comments. Should be good to go now!

**Comment by pbodnar** [comment]
>     2. I would just ... and add ` (#108)` to the end of the summary line.
(note: after looking in the commits history, I have added parentheses around the number)

@anderskaplan, maybe you have overlooked this part (which groups the commits together)? Apart from that, the commits seem perfect now. :)

**Review by pbodnar — APPROVED** [review]

**Comment by pbodnar** [comment]
Thank you!
